File: electron/src/scheduler.ts
/**
 * DeerFlow Electron - Task Scheduler
 *
 * Provides scheduled task execution for the agent platform:
 * - Cron-style scheduling (recurring tasks)
 * - One-time delayed execution
 * - Task persistence across app restarts
 * - Task categories: workflow, session, skill, system
 * - Execution history and retry logic
 * - Integration with EventBus for task events
 *
 * Uses pure Node.js timers (no external cron dependencies).
 */

import { EventEmitter } from "events";
import * as fs from "fs";
import * as path from "path";

// ============================================================
// Type Definitions
// ============================================================

export type TaskType = "workflow" | "session" | "skill" | "system" | "backup" | "cleanup";
export type TaskStatus = "pending" | "running" | "completed" | "failed" | "cancelled" | "skipped";

export interface ScheduledTask {
  id: string;
  name: string;
  description?: string;
  type: TaskType;
  status: TaskStatus;
  // Schedule config
  schedule: {
    type: "once" | "interval" | "cron";
    at?: string; // ISO timestamp for one-time
    intervalMs?: number; // For interval type
    cron?: string; // For cron type (simplified: "* * * * *")
  };
  // What to execute
  action: {
    handler: string; // e.g., "workflow:execute", "session:create", "skill:run"
    params: Record<string, any>;
  };
  // Execution config
  config: {
    enabled: boolean;
    maxRetries: number;
    retryDelayMs: number;
    timeoutMs: number;
    skipIfRunning: boolean;
  };
  // State
  lastRunAt?: string;
  lastRunResult?: "success" | "failed" | "timeout";
  lastRunError?: string;
  nextRunAt?: string;
  runCount: number;
  failCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface TaskExecution {
  id: string;
  taskId: string;
  startedAt: string;
  completedAt?: string;
  status: TaskStatus;
  result?: any;
  error?: string;
  durationMs?: number;
  retryCount: number;
}

export interface SchedulerConfig {
  maxConcurrentTasks: number;
  defaultTimeoutMs: number;
  defaultRetryCount: number;
  historyLimit: number;
  checkIntervalMs: number;
}

// ============================================================
// Default Configuration
// ============================================================

const DEFAULT_CONFIG: SchedulerConfig = {
  maxConcurrentTasks: 5,
  defaultTimeoutMs: 300000, // 5 minutes
  defaultRetryCount: 2,
  historyLimit: 1000,
  checkIntervalMs: 1000, // Check every second
};

// ============================================================
// Task Scheduler
// ============================================================

const SCHEDULER_DIR = "scheduler";
const TASKS_FILE = "tasks.json";
const HISTORY_FILE = "history.json";

export class TaskScheduler extends EventEmitter {
  private projectRoot: string;
  private schedulerDir: string;
  private tasks: Map<string, ScheduledTask> = new Map();
  private executions: Map<string, TaskExecution> = new Map();
  private timers: Map<string, NodeJS.Timeout> = new Map();
  private runningTasks: Set<string> = new Set();
  private config: SchedulerConfig;
  private checkTimer: NodeJS.Timeout | null = null;
  private dirty = false;

  constructor(projectRoot: string, config?: Partial<SchedulerConfig>) {
    super();
    this.projectRoot = projectRoot;
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.schedulerDir = path.join(projectRoot, ".deerflow", SCHEDULER_DIR);
    this.ensureDirectories();
    this.loadTasks();
    this.startScheduler();
  }

  // ============================================================
  // Task CRUD
  // ============================================================

  /**
   * Create a new scheduled task
   */
  createTask(
    task: Omit<ScheduledTask, "id" | "status" | "runCount" | "failCount" | "createdAt" | "updatedAt">
  ): ScheduledTask {
    const now = new Date().toISOString();
    const newTask: ScheduledTask = {
      ...task,
      id: this.generateId(),
      status: "pending",
      runCount: 0,
      failCount: 0,
      createdAt: now,
      updatedAt: now,
      config: {
        enabled: true,
        maxRetries: 2,
        retryDelayMs: 5000,
        timeoutMs: 300000,
        skipIfRunning: true,
        ...task.config,
      },
    };

    // Calculate next run time
    newTask.nextRunAt = this.calculateNextRun(newTask);

    this.tasks.set(newTask.id, newTask);
    this.dirty = true;
    this.saveTasks();

    // Schedule the task
    this.scheduleTask(newTask);

    this.emit("task:created", newTask);
    return newTask;
  }

  /**
   * Get a task by ID
   */
  getTask(id: string): ScheduledTask | null {
    return this.tasks.get(id) || null;
  }

  /**
   * List all tasks with optional filtering
   */
  listTasks(filter?: { type?: TaskType; status?: TaskStatus; enabled?: boolean }): ScheduledTask[] {
    let tasks = Array.from(this.tasks.values());

    if (filter?.type) {
      tasks = tasks.filter((t) => t.type === filter.type);
    }
    if (filter?.status) {
      tasks = tasks.filter((t) => t.status === filter.status);
    }
    if (filter?.enabled !== undefined) {
      tasks = tasks.filter((t) => t.config.enabled === filter.enabled);
    }

    return tasks.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
  }

  /**
   * Update a task
   */
  updateTask(id: string, updates: Partial<Omit<ScheduledTask, "id" | "createdAt">>): ScheduledTask | null {
    const task = this.tasks.get(id);
    if (!task) return null;

    Object.assign(task, updates, { updatedAt: new Date().toISOString() });

    // Recalculate next run if schedule changed
    if (updates.schedule) {
      task.nextRunAt = this.calculateNextRun(task);
    }

    this.dirty = true;
    this.saveTasks();

    // Reschedule
    this.unscheduleTask(id);
    this.scheduleTask(task);

    this.emit("task:updated", task);
    return task;
  }

  /**
   * Delete a task
   */
  deleteTask(id: string): boolean {
    const task = this.tasks.get(id);
    if (!task) return false;

    this.unscheduleTask(id);
    this.tasks.delete(id);
    this.dirty = true;
    this.saveTasks();

    this.emit("task:deleted", id);
    return true;
  }

  /**
   * Enable a task
   */
  enableTask(id: string): boolean {
    const task = this.tasks.get(id);
    if (!task) return false;

    task.config.enabled = true;
    task.updatedAt = new Date().toISOString();
    task.nextRunAt = this.calculateNextRun(task);

    this.dirty = true;
    this.saveTasks();
    this.scheduleTask(task);

    this.emit("task:enabled", task);
    return true;
  }

  /**
   * Disable a task
   */
  disableTask(id: string): boolean {
    const task = this.tasks.get(id);
    if (!task) return false;

    task.config.enabled = false;
    task.updatedAt = new Date().toISOString();
    task.nextRunAt = undefined;

    this.dirty = true;
    this.saveTasks();
    this.unscheduleTask(id);

    this.emit("task:disabled", task);
    return true;
  }

  // ============================================================
  // Task Execution
  // ============================================================

  /**
   * Run a task immediately (manual trigger)
   */
  async runTaskNow(id: string): Promise<TaskExecution | null> {
    const task = this.tasks.get(id);
    if (!task) return null;

    return this.executeTask(task);
  }

  /**
   * Execute a task
   */
  private async executeTask(task: ScheduledTask): Promise<TaskExecution> {
    const executionId = `exec-${Date.now()}-${Math.random().toString(36).substr(2, 6)}`;
    const startedAt = new Date().toISOString();

    const execution: TaskExecution = {
      id: executionId,
      taskId: task.id,
      startedAt,
      status: "running",
      retryCount: 0,
    };

    this.executions.set(executionId, execution);
    this.runningTasks.add(task.id);
    task.status = "running";

    this.emit("execution:started", execution, task);

    const startTime = Date.now();

    try {
      // Execute with timeout
      const result = await this.runWithTimeout(
        () => this.executeAction(task.action),
        task.config.timeoutMs || this.config.defaultTimeoutMs
      );

      execution.status = "completed";
      execution.result = result;
      execution.completedAt = new Date().toISOString();
      execution.durationMs = Date.now() - startTime;

      task.lastRunAt = execution.completedAt;
      task.lastRunResult = "success";
      task.lastRunError = undefined;
      task.runCount++;
      task.status = "pending";

      // Calculate next run for recurring tasks
      if (task.schedule.type !== "once") {
        task.nextRunAt = this.calculateNextRun(task);
      }

      this.emit("execution:completed", execution, task);
    } catch (err: any) {
      execution.status = "failed";
      execution.error = err.message;
      execution.completedAt = new Date().toISOString();
      execution.durationMs = Date.now() - startTime;

      task.lastRunAt = execution.completedAt;
      task.lastRunResult = "failed";
      task.lastRunError = err.message;
      task.failCount++;
      task.status = "pending";

      this.emit("execution:failed", execution, task, err);

      // Retry logic
      if (execution.retryCount < task.config.maxRetries) {
        execution.retryCount++;
        setTimeout(() => {
          this.executeTask(task);
        }, task.config.retryDelayMs);
      }
    } finally {
      this.runningTasks.delete(task.id);
      this.dirty = true;
      this.saveTasks();
      this.saveHistory();
    }

    return execution;
  }

  /**
   * Execute the task action
   */
  private async executeAction(action: ScheduledTask["action"]): Promise<any> {
    // Emit event for external handlers to process
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error("Task execution timeout"));
      }, this.config.defaultTimeoutMs);

      this.emit("action:execute", action, (result: any, error?: Error) => {
        clearTimeout(timeout);
        if (error) reject(error);
        else resolve(result);
      });

      // If no listener handles the action, resolve with placeholder
      setTimeout(() => {
        if (this.listenerCount("action:execute") === 0) {
          clearTimeout(timeout);
          resolve({ handled: false, action });
        }
      }, 100);
    });
  }

  /**
   * Run a function with timeout
   */
  private runWithTimeout<T>(fn: () => Promise<T>, timeoutMs: number): Promise<T> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`Task timed out after ${timeoutMs}ms`));
      }, timeoutMs);

      fn()
        .then((result) => {
          clearTimeout(timer);
          resolve(result);
        })
        .catch((err) => {
          clearTimeout(timer);
          reject(err);
        });
    });
  }

  // ============================================================
  // Scheduling Logic
  // ============================================================

  /**
   * Calculate next run time for a task
   */
  private calculateNextRun(task: ScheduledTask): string | undefined {
    if (!task.config.enabled) return undefined;

    const now = Date.now();

    switch (task.schedule.type) {
      case "once":
        if (task.schedule.at) {
          const atTime = new Date(task.schedule.at).getTime();
          return atTime > now ? task.schedule.at : undefined;
        }
        return undefined;

      case "interval":
        if (task.schedule.intervalMs) {
          const lastRun = task.lastRunAt ? new Date(task.lastRunAt).getTime() : now;
          const nextRun = lastRun + task.schedule.intervalMs;
          return new Date(Math.max(nextRun, now)).toISOString();
        }
        return undefined;

      case "cron": {
        // Simplified cron: interpret as minutes interval for now
        // Format: "*/5" = every 5 minutes
        const cron = task.schedule.cron || "*/60";
        const match = cron.match(/\*\/(\d+)/);
        if (match) {
          const minutes = parseInt(match[1], 10);
          const nextRun = now + minutes * 60 * 1000;
          return new Date(nextRun).toISOString();
        }
        return new Date(now + 3600000).toISOString(); // Default 1 hour
      }

      default:
        return undefined;
    }
  }

  /**
   * Schedule a task timer
   */
  private scheduleTask(task: ScheduledTask): void {
    if (!task.config.enabled || !task.nextRunAt) return;

    const nextRun = new Date(task.nextRunAt).getTime();
    const delay = Math.max(0, nextRun - Date.now());

    const timer = setTimeout(() => {
      this.timers.delete(task.id);
      if (task.config.skipIfRunning && this.runningTasks.has(task.id)) {
        task.status = "skipped";
        task.nextRunAt = this.calculateNextRun(task);
        this.scheduleTask(task);
        return;
      }
      this.executeTask(task);
    }, delay);

    this.timers.set(task.id, timer);
  }

  /**
   * Unschedule a task timer
   */
  private unscheduleTask(id: string): void {
    const timer = this.timers.get(id);
    if (timer) {
      clearTimeout(timer);
      this.timers.delete(id);
    }
  }

  /**
   * Start the scheduler loop
   */
  private startScheduler(): void {
    // Check for tasks that need to run every second
    this.checkTimer = setInterval(() => {
      this.checkTasks();
    }, this.config.checkIntervalMs);

    // Schedule all enabled tasks
    for (const task of this.tasks.values()) {
      if (task.config.enabled) {
        this.scheduleTask(task);
      }
    }
  }

  /**
   * Check and run due tasks
   */
  private checkTasks(): void {
    const now = Date.now();

    for (const task of this.tasks.values()) {
      if (!task.config.enabled || task.status === "running") continue;
      if (!task.nextRunAt) continue;

      const nextRun = new Date(task.nextRunAt).getTime();
      if (nextRun <= now) {
        // Reschedule if timer missed it
        this.unscheduleTask(task.id);
        this.executeTask(task);
      }
    }
  }

  // ============================================================
  // History
  // ============================================================

  /**
   * Get execution history for a task
   */
  getHistory(taskId?: string): TaskExecution[] {
    let history = Array.from(this.executions.values());
    if (taskId) {
      history = history.filter((e) => e.taskId === taskId);
    }
    return history.sort((a, b) => new Date(b.startedAt).getTime() - new Date(a.startedAt).getTime());
  }

  /**
   * Alias for getHistory - gets execution history
   */
  getExecutionHistory(taskId?: string): TaskExecution[] {
    return this.getHistory(taskId);
  }

  /**
   * Get recent executions
   */
  getRecentExecutions(limit: number = 50): TaskExecution[] {
    return this.getHistory().slice(0, limit);
  }

  /**
   * Clear old history
   */
  clearOldHistory(keepCount: number = this.config.historyLimit): void {
    const all = this.getHistory();
    if (all.length <= keepCount) return;

    const toKeep = all.slice(0, keepCount);
    this.executions.clear();
    for (const exec of toKeep) {
      this.executions.set(exec.id, exec);
    }
    this.dirty = true;
    this.saveHistory();
  }

  // ============================================================
  // Statistics
  // ============================================================

  getStats(): {
    totalTasks: number;
    enabledTasks: number;
    runningTasks: number;
    byType: Record<string, number>;
    totalExecutions: number;
    successfulExecutions: number;
    failedExecutions: number;
  } {
    const tasks = Array.from(this.tasks.values());
    const executions = Array.from(this.executions.values());
    const byType: Record<string, number> = {};

    for (const t of tasks) {
      byType[t.type] = (byType[t.type] || 0) + 1;
    }

    return {
      totalTasks: tasks.length,
      enabledTasks: tasks.filter((t) => t.config.enabled).length,
      runningTasks: tasks.filter((t) => t.status === "running").length,
      byType,
      totalExecutions: executions.length,
      successfulExecutions: executions.filter((e) => e.status === "completed").length,
      failedExecutions: executions.filter((e) => e.status === "failed").length,
    };
  }

  // ============================================================
  // Persistence
  // ============================================================

  private ensureDirectories(): void {
    if (!fs.existsSync(this.schedulerDir)) {
      fs.mkdirSync(this.schedulerDir, { recursive: true });
    }
  }

  private getTasksPath(): string {
    return path.join(this.schedulerDir, TASKS_FILE);
  }

  private getHistoryPath(): string {
    return path.join(this.schedulerDir, HISTORY_FILE);
  }

  private loadTasks(): void {
    try {
      const tasksPath = this.getTasksPath();
      if (!fs.existsSync(tasksPath)) return;

      const data = JSON.parse(fs.readFileSync(tasksPath, "utf-8"));
      for (const task of data.tasks || []) {
        this.tasks.set(task.id, task);
      }
      console.log(`[Scheduler] Loaded ${this.tasks.size} tasks`);
    } catch (err) {
      console.warn("[Scheduler] Failed to load tasks:", err);
    }
  }

  private saveTasks(): void {
    if (!this.dirty) return;
    try {
      const tasksPath = this.getTasksPath();
      const data = {
        updatedAt: new Date().toISOString(),
        tasks: Array.from(this.tasks.values()),
      };
      fs.writeFileSync(tasksPath, JSON.stringify(data, null, 2), "utf-8");
    } catch (err) {
      console.warn("[Scheduler] Failed to save tasks:", err);
    }
  }

  private saveHistory(): void {
    try {
      const historyPath = this.getHistoryPath();
      const data = {
        updatedAt: new Date().toISOString(),
        executions: Array.from(this.executions.values()).slice(-this.config.historyLimit),
      };
      fs.writeFileSync(historyPath, JSON.stringify(data, null, 2), "utf-8");
    } catch (err) {
      console.warn("[Scheduler] Failed to save history:", err);
    }
  }

  // ============================================================
  // Cleanup
  // ============================================================

  destroy(): void {
    if (this.checkTimer) {
      clearInterval(this.checkTimer);
      this.checkTimer = null;
    }

    for (const timer of this.timers.values()) {
      clearTimeout(timer);
    }
    this.timers.clear();

    this.saveTasks();
    this.saveHistory();
    this.removeAllListeners();
  }

  private generateId(): string {
    return `task-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }
}

File: backend/app/gateway/routers/scheduler.py
"""Task Scheduler REST API Router.

Provides CRUD endpoints for scheduled tasks, enabling browser-mode access
to the task scheduling system. Uses JSON-file persistence.

Includes a background asyncio loop that evaluates interval/once tasks and
executes them when due. Handlers can route to specific modules (e.g. backup).

Endpoints
---------
GET    /api/electron/scheduler/tasks       – list tasks (optional filters)
POST   /api/electron/scheduler/tasks       – create task
GET    /api/electron/scheduler/tasks/{id}  – get task detail
PATCH  /api/electron/scheduler/tasks/{id}  – update task
DELETE /api/electron/scheduler/tasks/{id}  – delete task
POST   /api/electron/scheduler/tasks/{id}/enable  – enable task
POST   /api/electron/scheduler/tasks/{id}/disable – disable task
POST   /api/electron/scheduler/tasks/{id}/run     – trigger execution
GET    /api/electron/scheduler/history     – execution history
GET    /api/electron/scheduler/stats       – scheduler statistics
POST   /api/electron/scheduler/loop/start  – start background loop
POST   /api/electron/scheduler/loop/stop   – stop background loop
GET    /api/electron/scheduler/loop/status – loop status
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
import uuid
from datetime import datetime, timezone, timedelta
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from deerflow.config.paths import get_paths

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/electron/scheduler", tags=["scheduler"])

_lock = asyncio.Lock()
_tasks: dict[str, dict] = {}
_history: list[dict] = []
_persistence_path: Optional[Path] = None

# ── Background loop state ─────────────────────────────────────────────
_loop_task: asyncio.Task | None = None
_loop_running = False
_loop_interval_s = 15
_loop_last_check: float | None = None

# ── Pydantic models ───────────────────────────────────────────────────


class TaskSchedule(BaseModel):
    type: str = Field(..., description="once|interval|cron")
    intervalMs: int | None = Field(default=None, ge=1000, alias="intervalMs")
    cron: str | None = None
    at: str | None = None  # ISO timestamp for one-time

    model_config = {"populate_by_name": True}


class TaskAction(BaseModel):
    handler: str = Field(default="system:dummy")
    params: dict = Field(default_factory=dict)


class TaskConfig(BaseModel):
    enabled: bool = Field(default=True)
    maxRetries: int = Field(default=3, ge=0, alias="maxRetries")
    retryDelayMs: int = Field(default=5000, ge=1000, alias="retryDelayMs")
    timeoutMs: int = Field(default=300000, alias="timeoutMs")
    skipIfRunning: bool = Field(default=True, alias="skipIfRunning")

    model_config = {"populate_by_name": True}


class CreateTaskRequest(BaseModel):
    name: str = Field(..., min_length=1)
    description: str = Field(default="")
    type: str = Field(default="system", description="workflow|session|skill|system|backup|cleanup")
    schedule: TaskSchedule = Field(...)
    action: TaskAction = Field(default_factory=TaskAction)
    config: TaskConfig = Field(default_factory=TaskConfig)


class UpdateTaskRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    type: str | None = None
    schedule: TaskSchedule | None = None
    action: TaskAction | None = None
    config: TaskConfig | None = None


# ── persistence helpers ───────────────────────────────────────────────


def _get_persistence_path() -> Path:
    global _persistence_path
    if _persistence_path is None:
        try:
            base = get_paths().base_dir
        except Exception:
            base = Path(".")
        _persistence_path = base / "scheduled_tasks.json"
    return _persistence_path


async def _load_state() -> None:
    path = _get_persistence_path()
    if not path.exists():
        return
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        logger.exception("Failed to read scheduled_tasks.json")
        return
    for t in data.get("tasks", []):
        if isinstance(t, dict) and t.get("id"):
            _tasks[t["id"]] = t
    history_raw = data.get("history", [])
    _history.clear()
    _history.extend(history_raw)
    logger.info("Loaded %d tasks and %d history entries", len(_tasks), len(_history))


async def _save_state() -> None:
    path = _get_persistence_path()
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        data = {
            "tasks": list(_tasks.values()),
            "history": _history[-500:],  # Keep last 500 entries
            "updatedAt": time.time(),
        }
        path.write_text(json.dumps(data, indent=2), encoding="utf-8")
    except Exception:
        logger.exception("Failed to persist scheduled tasks")


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


async def get_tasks() -> list[dict]:
    """Return a snapshot of all tasks as a list (safe for cross-module import)."""
    return list(_tasks.values())


# ── Task CRUD endpoints ───────────────────────────────────────────────


@router.post("/tasks")
async def create_task(req: CreateTaskRequest):
    async with _lock:
        tid = str(uuid.uuid4())
        now = _now_iso()
        task = {
            "id": tid,
            "name": req.name,
            "description": req.description,
            "type": req.type,
            "status": "pending",
            "schedule": req.schedule.model_dump(by_alias=True) if hasattr(req.schedule, "model_dump") else req.schedule.dict(),
            "action": req.action.model_dump() if hasattr(req.action, "model_dump") else req.action.dict(),
            "config": req.config.model_dump(by_alias=True) if hasattr(req.config, "model_dump") else req.config.dict(),
            "runCount": 0,
            "failCount": 0,
            "createdAt": now,
            "updatedAt": now,
        }
        _tasks[tid] = task
        await _save_state()
        return task


@router.get("/tasks")
async def list_tasks(
    type: str | None = Query(default=None),
    status: str | None = Query(default=None),
    enabled: bool | None = Query(default=None),
):
    results = list(_tasks.values())
    if type:
        results = [t for t in results if t.get("type") == type]
    if status:
        results = [t for t in results if t.get("status") == status]
    if enabled is not None:
        results = [t for t in results if t.get("config", {}).get("enabled") == enabled]
    return results


@router.get("/tasks/{task_id}")
async def get_task(task_id: str):
    t = _tasks.get(task_id)
    if t is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return t


@router.patch("/tasks/{task_id}")
async def update_task(task_id: str, req: UpdateTaskRequest):
    async with _lock:
        t = _tasks.get(task_id)
        if t is None:
            raise HTTPException(status_code=404, detail="Task not found")
        if req.name is not None:
            t["name"] = req.name
        if req.description is not None:
            t["description"] = req.description
        if req.type is not None:
            t["type"] = req.type
        if req.schedule is not None:
            t["schedule"] = req.schedule.model_dump(by_alias=True) if hasattr(req.schedule, "model_dump") else req.schedule.dict()
        if req.action is not None:
            t["action"] = req.action.model_dump() if hasattr(req.action, "model_dump") else req.action.dict()
        if req.config is not None:
            t["config"] = req.config.model_dump(by_alias=True) if hasattr(req.config, "model_dump") else req.config.dict()
        t["updatedAt"] = _now_iso()
        await _save_state()
        return t


@router.delete("/tasks/{task_id}")
async def delete_task(task_id: str):
    async with _lock:
        if task_id not in _tasks:
            raise HTTPException(status_code=404, detail="Task not found")
        del _tasks[task_id]
        await _save_state()
        return {"success": True}


@router.post("/tasks/{task_id}/enable")
async def enable_task(task_id: str):
    async with _lock:
        t = _tasks.get(task_id)
        if t is None:
            raise HTTPException(status_code=404, detail="Task not found")
        t.setdefault("config", {})["enabled"] = True
        t["updatedAt"] = _now_iso()
        await _save_state()
        return {"success": True}


@router.post("/tasks/{task_id}/disable")
async def disable_task(task_id: str):
    async with _lock:
        t = _tasks.get(task_id)
        if t is None:
            raise HTTPException(status_code=404, detail="Task not found")
        t.setdefault("config", {})["enabled"] = False
        t["updatedAt"] = _now_iso()
        await _save_state()
        return {"success": True}


@router.post("/tasks/{task_id}/run")
async def run_task_now(task_id: str):
    t = _tasks.get(task_id)
    if t is None:
        raise HTTPException(status_code=404, detail="Task not found")
    exec_id = str(uuid.uuid4())
    now = _now_iso()
    execution = {
        "id": exec_id,
        "taskId": task_id,
        "startedAt": now,
        "status": "completed",  # Placeholder — real execution happens in Electron
        "retryCount": 0,
        "durationMs": 0,
    }
    async with _lock:
        t["runCount"] = t.get("runCount", 0) + 1
        t["lastRunAt"] = now
        t["lastRunResult"] = "success"
        t["updatedAt"] = now
        _history.append(execution)
        await _save_state()
        return execution


# ── History & Stats endpoints ─────────────────────────────────────────


@router.get("/history")
async def get_history(
    taskId: str | None = Query(default=None, alias="taskId"),
):
    results = _history
    if taskId:
        results = [h for h in results if h.get("taskId") == taskId]
    return results


@router.get("/stats")
async def get_stats():
    total = len(_tasks)
    enabled_count = sum(1 for t in _tasks.values() if t.get("config", {}).get("enabled"))
    running_count = sum(1 for t in _tasks.values() if t.get("status") == "running")
    by_type: dict[str, int] = {}
    for t in _tasks.values():
        tp = t.get("type", "unknown")
        by_type[tp] = by_type.get(tp, 0) + 1
    total_exec = sum(t.get("runCount", 0) for t in _tasks.values())
    failed = sum(t.get("failCount", 0) for t in _tasks.values())
    successful = total_exec - failed

    disabled_count = total - enabled_count
    from datetime import datetime, timezone, timedelta
    today_start = datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0)
    executions_today = sum(
        1 for h in _history
        if datetime.fromisoformat(h.get("startedAt", "")).replace(tzinfo=timezone.utc) >= today_start
    )

    return {
        "totalTasks": total,
        "enabledTasks": enabled_count,
        "disabledTasks": disabled_count,
        "runningTasks": running_count,
        "byType": by_type,
        "totalExecutions": total_exec,
        "successfulExecutions": successful,
        "failedExecutions": failed,
        "executionsToday": executions_today,
    }


# ── Background loop ───────────────────────────────────────────────────


async def _execute_task(task: dict) -> None:
    """Execute a single task by dispatching to the appropriate handler."""
    handler = task.get("action", {}).get("handler", "system:dummy")
    tid = task["id"]
    now = _now_iso()

    execution = {
        "id": str(uuid.uuid4()),
        "taskId": tid,
        "startedAt": now,
        "status": "completed",
        "retryCount": 0,
        "durationMs": 0,
        "error": None,
    }

    try:
        if handler == "system:backup":
            # Delegate to backup module
            try:
                from app.gateway.routers.backup import create_backup_async
                result = await create_backup_async()
                execution["result"] = result
                logger.info("Auto-backup completed for task %s", tid)
            except ImportError:
                execution["status"] = "failed"
                execution["error"] = "Backup module not available"
                logger.warning("Backup module not importable for task %s", tid)
        else:
            # Generic handler: log and record
            logger.info("Executing task %s (handler=%s)", tid, handler)

    except Exception as exc:
        execution["status"] = "failed"
        execution["error"] = str(exc)
        logger.exception("Task %s execution failed", tid)

    execution["completedAt"] = _now_iso()

    async with _lock:
        task["runCount"] = task.get("runCount", 0) + 1
        task["lastRunAt"] = now
        if execution["status"] == "failed":
            task["failCount"] = task.get("failCount", 0) + 1
            task["lastRunResult"] = "failure"
            task["lastRunError"] = execution.get("error")
        else:
            task["lastRunResult"] = "success"
        task["updatedAt"] = now

        # Schedule next run for interval tasks
        sched = task.get("schedule", {})
        if sched.get("type") == "interval":
            interval_ms = sched.get("intervalMs", 3600000)
            task["nextRunAt"] = (datetime.now(timezone.utc) + timedelta(milliseconds=interval_ms)).isoformat()

        _history.append(execution)
        await _save_state()


async def _scheduler_loop() -> None:
    """Periodically check for due tasks and execute them."""
    global _loop_running, _loop_last_check
    logger.info("Scheduler background loop started (interval=%ds)", _loop_interval_s)
    while _loop_running:
        try:
            _loop_last_check = time.time()
            now = datetime.now(timezone.utc)
            due_tasks = []

            async with _lock:
                for t in _tasks.values():
                    if not t.get("config", {}).get("enabled"):
                        continue
                    sched = t.get("schedule", {})
                    stype = sched.get("type")

                    if stype == "once":
                        at_str = sched.get("at")
                        if at_str:
                            try:
                                at_dt = datetime.fromisoformat(at_str)
                                if at_dt.tzinfo is None:
                                    at_dt = at_dt.replace(tzinfo=timezone.utc)
                                if at_dt <= now and (not t.get("lastRunAt") or t.get("status") != "completed"):
                                    due_tasks.append(t)
                            except ValueError:
                                pass

                    elif stype == "interval":
                        next_at_str = t.get("nextRunAt")
                        if not next_at_str:
                            # First run: use creation time + interval
                            interval_ms = sched.get("intervalMs", 3600000)
                            created = t.get("createdAt", _now_iso())
                            try:
                                created_dt = datetime.fromisoformat(created)
                                if created_dt.tzinfo is None:
                                    created_dt = created_dt.replace(tzinfo=timezone.utc)
                                next_at = created_dt + timedelta(milliseconds=interval_ms)
                            except ValueError:
                                next_at = now + timedelta(milliseconds=interval_ms)
                        else:
                            try:
                                next_at = datetime.fromisoformat(next_at_str)
                                if next_at.tzinfo is None:
                                    next_at = next_at.replace(tzinfo=timezone.utc)
                            except ValueError:
                                next_at = now

                        if next_at <= now:
                            due_tasks.append(t)

            # Execute due tasks
            for task in due_tasks:
                if not _loop_running:
                    break
                await _execute_task(task)

        except Exception:
            logger.exception("Scheduler loop error (will retry next cycle)")

        # Sleep until next check
        try:
            await asyncio.sleep(_loop_interval_s)
        except asyncio.CancelledError:
            break

    _loop_running = False
    logger.info("Scheduler background loop stopped")


async def start_scheduler_loop() -> None:
    """Start the background scheduler loop (called from app lifespan)."""
    global _loop_task, _loop_running
    if _loop_running:
        logger.info("Scheduler loop already running")
        return
    _loop_running = True
    _loop_task = asyncio.create_task(_scheduler_loop())


async def stop_scheduler_loop() -> None:
    """Stop the background scheduler loop (called from app lifespan)."""
    global _loop_task, _loop_running
    _loop_running = False
    if _loop_task:
        _loop_task.cancel()
        try:
            await _loop_task
        except asyncio.CancelledError:
            pass
        _loop_task = None


# ── Loop control endpoints ────────────────────────────────────────────


@router.post("/loop/start")
async def loop_start():
    await start_scheduler_loop()
    return {"running": True}


@router.post("/loop/stop")
async def loop_stop():
    await stop_scheduler_loop()
    return {"running": False}


@router.get("/loop/status")
async def loop_status():
    return {
        "running": _loop_running,
        "lastCheck": _loop_last_check,
        "intervalSeconds": _loop_interval_s,
        "taskCount": len(_tasks),
        "historyCount": len(_history),
    }
